File: src/protobufCodec.ts
/**
 * protobufCodec.ts
 *
 * 纯 TypeScript protobuf wire format 编解码器。
 * 基于 2026-03-24 对 state.vscdb 中 CascadeTrajectorySummary 的逆向分析。
 *
 * state.vscdb 中 `antigravityUnifiedStateSync.trajectorySummaries` 的存储格式：
 *   value = base64(outer_protobuf)
 *
 * outer_protobuf 结构:
 *   repeated field 1 (map_entry) {
 *     field 1: cascadeId (string)        — map key
 *     field 2: wrapper_message {         — map value
 *       field 1: base64(inner_protobuf)  — double encoding!
 *     }
 *   }
 *
 * inner_protobuf (CascadeTrajectorySummary):
 *   field 1: summary (string)
 *   field 2: stepCount (varint)
 *   field 3: lastModifiedTime (Timestamp: {1:seconds, 2:nanos})
 *   field 4: trajectoryId (string)
 *   field 5: status (varint/enum)
 *   field 7: createdTime (Timestamp)
 *   field 9: workspaces (repeated message: {1:uri, 2:rootUri, 3:repo})
 *   field 10: lastUserInputTime (Timestamp)
 */

import type { TrajectorySummary } from './lsClient';

// ─── Wire format primitives ───

export function encodeVarint(value: number): Buffer {
  const bytes: number[] = [];
  let v = value >>> 0; // ensure unsigned for small values
  // Handle large numbers (> 32-bit) via BigInt-style encoding
  if (value > 0x7FFFFFFF) {
    // Use string conversion for large epoch seconds
    let big = BigInt(value);
    while (big > 0x7Fn) {
      bytes.push(Number(big & 0x7Fn) | 0x80);
      big >>= 7n;
    }
    bytes.push(Number(big & 0x7Fn));
    return Buffer.from(bytes);
  }
  while (v > 0x7F) {
    bytes.push((v & 0x7F) | 0x80);
    v >>>= 7;
  }
  bytes.push(v & 0x7F);
  return Buffer.from(bytes);
}

export function encodeLengthDelimited(fieldNumber: number, data: Buffer): Buffer {
  const tag = encodeVarint((fieldNumber << 3) | 2);
  const len = encodeVarint(data.length);
  return Buffer.concat([tag, len, data]);
}

export function encodeString(fieldNumber: number, value: string): Buffer {
  return encodeLengthDelimited(fieldNumber, Buffer.from(value, 'utf-8'));
}

export function encodeVarintField(fieldNumber: number, value: number): Buffer {
  const tag = encodeVarint((fieldNumber << 3) | 0);
  return Buffer.concat([tag, encodeVarint(value)]);
}

export function encodeTimestamp(fieldNumber: number, seconds: number, nanos: number): Buffer {
  let inner = encodeVarintField(1, seconds);
  if (nanos) {
    inner = Buffer.concat([inner, encodeVarintField(2, nanos)]);
  }
  return encodeLengthDelimited(fieldNumber, inner);
}

// ─── ISO 8601 timestamp parsing ───

export function parseIsoTimestamp(ts: string): { seconds: number; nanos: number } | null {
  if (!ts) return null;
  try {
    let cleaned = ts.replace(/Z$/, '');
    let nanos = 0;
    const dotIdx = cleaned.indexOf('.');
    if (dotIdx >= 0) {
      const frac = cleaned.substring(dotIdx + 1).padEnd(9, '0').substring(0, 9);
      nanos = parseInt(frac, 10);
      cleaned = cleaned.substring(0, dotIdx);
    }
    const dt = new Date(cleaned + 'Z');
    const seconds = Math.floor(dt.getTime() / 1000);
    return { seconds, nanos };
  } catch {
    return null;
  }
}

// ─── Status enum mapping ───

const STATUS_MAP: Record<string, number> = {
  CASCADE_RUN_STATUS_UNSPECIFIED: 0,
  CASCADE_RUN_STATUS_IDLE: 1,
  CASCADE_RUN_STATUS_RUNNING: 2,
  CASCADE_RUN_STATUS_STREAMING: 3,
  CASCADE_RUN_STATUS_DONE: 4,
};

// ─── CascadeTrajectorySummary encoder ───

/**
 * Encode a CascadeTrajectorySummary from LS API JSON to protobuf binary.
 */
export function encodeCascadeSummary(summary: TrajectorySummary): Buffer {
  const parts: Buffer[] = [];

  // f1: summary (string)
  if (summary.summary) {
    parts.push(encodeString(1, summary.summary));
  }

  // f2: stepCount (varint)
  if (summary.stepCount) {
    parts.push(encodeVarintField(2, summary.stepCount));
  }

  // f3: lastModifiedTime (Timestamp)
  const lmt = parseIsoTimestamp(summary.lastModifiedTime);
  if (lmt) {
    parts.push(encodeTimestamp(3, lmt.seconds, lmt.nanos));
  }

  // f4: trajectoryId (string)
  if (summary.trajectoryId) {
    parts.push(encodeString(4, summary.trajectoryId));
  }

  // f5: status (varint/enum)
  const statusVal = STATUS_MAP[summary.status] ?? 1;
  if (statusVal) {
    parts.push(encodeVarintField(5, statusVal));
  }

  // f7: createdTime (Timestamp)
  const ct = parseIsoTimestamp(summary.createdTime);
  if (ct) {
    parts.push(encodeTimestamp(7, ct.seconds, ct.nanos));
  }

  // f9: workspaces (repeated message: {1:uri, 2:rootUri, 3:repo})
  if (summary.workspaces) {
    for (const ws of summary.workspaces) {
      const wsParts: Buffer[] = [];
      const uri = ws.workspaceFolderAbsoluteUri || '';
      if (uri) {
        wsParts.push(encodeString(1, uri));
        wsParts.push(encodeString(2, 'file:///'));
        wsParts.push(encodeString(3, ''));
      }
      if (wsParts.length > 0) {
        parts.push(encodeLengthDelimited(9, Buffer.concat(wsParts)));
      }
    }
  }

  // f10: lastUserInputTime (Timestamp)
  const luit = parseIsoTimestamp((summary as any).lastUserInputTime);
  if (luit) {
    parts.push(encodeTimestamp(10, luit.seconds, luit.nanos));
  }

  return Buffer.concat(parts);
}

// ─── Map entry builder ───

/**
 * Build a single map entry for the outer protobuf.
 *
 * Wire format:
 *   field 1 (LEN): map_entry {
 *     field 1 (LEN): cascadeId (string)
 *     field 2 (LEN): wrapper {
 *       field 1 (LEN): base64(summary_proto)    ← double encoding
 *     }
 *   }
 */
export function buildMapEntry(cascadeId: string, summaryProto: Buffer): Buffer {
  const b64Value = summaryProto.toString('base64');
  const wrapper = encodeString(1, b64Value);
  const entry = Buffer.concat([
    encodeString(1, cascadeId),
    encodeLengthDelimited(2, wrapper),
  ]);
  return encodeLengthDelimited(1, entry);
}

// ─── Parsing existing protobuf ───

/** Read a varint from buffer at pos, return [value, newPos] */
function readVarint(buf: Buffer, pos: number): [number, number] {
  let result = 0;
  let shift = 0;
  while (pos < buf.length) {
    const b = buf[pos++];
    result |= (b & 0x7F) << shift;
    shift += 7;
    if (!(b & 0x80)) break;
    // For large varints (> 28 bits), switch to safe math
    if (shift >= 28) {
      let big = BigInt(result);
      while (pos < buf.length) {
        const b2 = buf[pos++];
        big |= BigInt(b2 & 0x7F) << BigInt(shift);
        shift += 7;
        if (!(b2 & 0x80)) break;
      }
      return [Number(big), pos];
    }
  }
  return [result >>> 0, pos];
}

/**
 * Parse the outer protobuf to extract existing cascade IDs.
 */
export function parseExistingEntryIds(rawProto: Buffer): Set<string> {
  const ids = new Set<string>();
  let pos = 0;

  while (pos < rawProto.length) {
    const [tag, pos2] = readVarint(rawProto, pos);
    pos = pos2;
    const fieldNum = tag >>> 3;
    const wireType = tag & 0x07;

    if (wireType === 2) {
      const [len, pos3] = readVarint(rawProto, pos);
      pos = pos3;
      const entryData = rawProto.subarray(pos, pos + len);
      pos += len;

      if (fieldNum === 1) {
        // Parse map entry → extract key (field 1)
        let epos = 0;
        while (epos < entryData.length) {
          const [eTag, epos2] = readVarint(entryData, epos);
          epos = epos2;
          const eFn = eTag >>> 3;
          const eWt = eTag & 0x07;

          if (eWt === 2) {
            const [eLen, epos3] = readVarint(entryData, epos);
            epos = epos3;
            const eData = entryData.subarray(epos, epos + eLen);
            epos += eLen;
            if (eFn === 1) {
              ids.add(eData.toString('utf-8'));
            }
          } else {
            break; // unexpected wire type
          }
        }
      }
    } else {
      break; // unexpected top-level wire type
    }
  }

  return ids;
}

// ─── Full index builder ───

/**
 * Build a complete inject payload by appending new entries to existing protobuf.
 *
 * Protobuf repeated fields can be simply concatenated — this is safe
 * and preserves the exact binary format of existing entries.
 *
 * @param existingRaw - existing protobuf bytes from state.vscdb (preserved as-is)
 * @param newSummaries - map of cascadeId → TrajectorySummary to add
 * @returns combined protobuf bytes
 */
export function buildInjectPayload(
  existingRaw: Buffer | null,
  newSummaries: Map<string, TrajectorySummary>
): Buffer {
  const existingIds = existingRaw ? parseExistingEntryIds(existingRaw) : new Set<string>();
  const parts: Buffer[] = [];

  // Keep existing data as-is
  if (existingRaw && existingRaw.length > 0) {
    parts.push(existingRaw);
  }

  // Append new entries
  let added = 0;
  for (const [cascadeId, summary] of newSummaries) {
    if (existingIds.has(cascadeId)) continue;
    const summaryProto = encodeCascadeSummary(summary);
    parts.push(buildMapEntry(cascadeId, summaryProto));
    added++;
  }

  return Buffer.concat(parts);
}

File: Guardian/Sources/Services/ProtobufCodec.swift
import Foundation

/// 轻量 protobuf wire format 编解码器
/// 复现 AG 的 trajectorySummaries 存储格式
enum ProtobufCodec {

    // MARK: - Varint

    static func encodeVarint(_ value: UInt64) -> Data {
        var v = value
        var result = Data()
        while v > 0x7F {
            result.append(UInt8(v & 0x7F) | 0x80)
            v >>= 7
        }
        result.append(UInt8(v))
        return result
    }

    static func decodeVarint(from data: Data, offset: inout Int) -> UInt64? {
        var result: UInt64 = 0
        var shift: UInt64 = 0
        while offset < data.count {
            let byte = data[offset]
            offset += 1
            result |= UInt64(byte & 0x7F) << shift
            if byte & 0x80 == 0 { return result }
            shift += 7
            if shift > 63 { return nil }
        }
        return nil
    }

    // MARK: - Tag

    static func makeTag(fieldNumber: Int, wireType: Int) -> Data {
        encodeVarint(UInt64(fieldNumber << 3 | wireType))
    }

    // MARK: - Length-delimited field

    static func lengthDelimited(fieldNumber: Int, data: Data) -> Data {
        var result = makeTag(fieldNumber: fieldNumber, wireType: 2)
        result.append(encodeVarint(UInt64(data.count)))
        result.append(data)
        return result
    }

    static func stringField(fieldNumber: Int, value: String) -> Data {
        lengthDelimited(fieldNumber: fieldNumber, data: Data(value.utf8))
    }

    static func varintField(fieldNumber: Int, value: UInt64) -> Data {
        var result = makeTag(fieldNumber: fieldNumber, wireType: 0)
        result.append(encodeVarint(value))
        return result
    }

    // MARK: - Timestamp (Google Protobuf Timestamp)

    static func timestampField(fieldNumber: Int, date: Date) -> Data {
        let seconds = Int64(date.timeIntervalSince1970)
        let nanos = Int32((date.timeIntervalSince1970 - Double(seconds)) * 1_000_000_000)
        var inner = Data()
        if seconds != 0 {
            inner.append(makeTag(fieldNumber: 1, wireType: 0))
            inner.append(encodeVarint(UInt64(bitPattern: Int64(seconds))))
        }
        if nanos != 0 {
            inner.append(makeTag(fieldNumber: 2, wireType: 0))
            inner.append(encodeVarint(UInt64(bitPattern: Int64(nanos))))
        }
        return lengthDelimited(fieldNumber: fieldNumber, data: inner)
    }

    // MARK: - Map Entry

    /// 构造 CascadeSummary message
    static func encodeCascadeSummary(
        cascadeId: String,
        title: String,
        createdAt: Date,
        updatedAt: Date,
        stepCount: Int
    ) -> Data {
        var msg = Data()
        msg.append(stringField(fieldNumber: 1, value: cascadeId))      // cascade_id
        msg.append(stringField(fieldNumber: 2, value: title))          // title
        msg.append(timestampField(fieldNumber: 3, date: createdAt))    // created_time
        msg.append(timestampField(fieldNumber: 4, date: updatedAt))    // updated_time
        msg.append(varintField(fieldNumber: 5, value: UInt64(stepCount)))  // step_count
        return msg
    }

    /// 构造 map<string, CascadeSummary> entry
    static func buildMapEntry(key: String, value: Data) -> Data {
        var entry = Data()
        entry.append(stringField(fieldNumber: 1, value: key))    // map key
        entry.append(lengthDelimited(fieldNumber: 2, data: value))  // map value
        return lengthDelimited(fieldNumber: 1, data: entry)       // field 1 of container
    }

    // MARK: - Payload

    static func buildPayload(
        selectedIds: Set<String>,
        summaries: [String: [String: Any]]
    ) -> Data {
        let isoFormatter = ISO8601DateFormatter()
        var payload = Data()

        for id in selectedIds {
            let s = summaries[id]
            let title = (s?["title"] as? String) ?? ""
            let created = (s?["createdTime"] as? String).flatMap { isoFormatter.date(from: $0) } ?? Date()
            let updated = (s?["lastModifiedTime"] as? String).flatMap { isoFormatter.date(from: $0) } ?? Date()
            let steps = (s?["stepCount"] as? Int) ?? 1

            let summary = encodeCascadeSummary(
                cascadeId: id,
                title: title,
                createdAt: created,
                updatedAt: updated,
                stepCount: steps
            )
            payload.append(buildMapEntry(key: id, value: summary))
        }

        return payload
    }

    // MARK: - Parse existing IDs

    /// 从现有 protobuf 数据中提取所有 cascadeId
    static func parseEntryIds(from data: Data) -> Set<String> {
        var ids = Set<String>()
        var offset = 0

        while offset < data.count {
            // 读 tag
            guard let tag = decodeVarint(from: data, offset: &offset) else { break }
            let wireType = Int(tag & 0x7)

            if wireType == 2 {
                // length-delimited
                guard let length = decodeVarint(from: data, offset: &offset) else { break }
                let end = offset + Int(length)
                guard end <= data.count else { break }

                // 尝试解析 map entry：field 1 = key (string)
                let entryData = data[offset..<end]
                if let key = extractMapKey(from: Data(entryData)) {
                    ids.insert(key)
                }
                offset = end
            } else if wireType == 0 {
                // varint: skip
                _ = decodeVarint(from: data, offset: &offset)
            } else {
                // 无法解析，退出
                break
            }
        }

        return ids
    }

    private static func extractMapKey(from data: Data) -> String? {
        var offset = 0
        guard let tag = decodeVarint(from: data, offset: &offset) else { return nil }
        let fieldNumber = Int(tag >> 3)
        let wireType = Int(tag & 0x7)

        guard fieldNumber == 1, wireType == 2 else { return nil }
        guard let length = decodeVarint(from: data, offset: &offset) else { return nil }
        let end = offset + Int(length)
        guard end <= data.count else { return nil }

        return String(data: data[offset..<end], encoding: .utf8)
    }
}
